pandas/io/common.py

```diff
@@ -165,11 +165,80 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
+            cycle through a file in S3: after reaching the end of the file,
+            boto will close the file. Then the next call to `read` or `next` will
+            re-open the file and start reading from the beginning.
+
+            Also adds a `readline` function which will split the returned
+            values by the `\n` character.
+            """
+            def __init__(self, *args, **kwargs):
+                encoding = kwargs.pop("encoding", None)  # Python 2 compat
+                super(OnceThroughKey, self).__init__(*args, **kwargs)
+                self.finished_read = False  # Add a flag to mark the end of the read.
+                self.buffer = ""
+                self.lines = []
+                if encoding is None and compat.PY3:
+                    encoding = "utf-8"
+                self.encoding = encoding
+                self.lines = []
+
+            def next(self):
+                return self.readline()
+
+            __next__ = next
+
+            def read(self, *args, **kwargs):
+                if self.finished_read:
+                    return b'' if compat.PY3 else ''
+                return super(OnceThroughKey, self).read(*args, **kwargs)
+
+            def close(self, *args, **kwargs):
+                self.finished_read = True
+                return super(OnceThroughKey, self).close(*args, **kwargs)
+
+            def seekable(self):
+                """Needed for reading by bz2"""
+                return False
+
+            def readline(self):
+                """Split the contents of the Key by '\n' characters."""
+                if self.lines:
+                    retval = self.lines[0]
+                    self.lines = self.lines[1:]
+                    return retval
+                if self.finished_read:
+                    if self.buffer:
+                        retval, self.buffer = self.buffer, ""
+                        return retval
+                    else:
+                        raise StopIteration
+
+                if self.encoding:
+                    self.buffer = "{}{}".format(self.buffer, self.read(8192).decode(self.encoding))
+                else:
+                    self.buffer = "{}{}".format(self.buffer, self.read(8192))
+
+                split_buffer = self.buffer.split("\n")
+                self.lines.extend(split_buffer[:-1])
+                self.buffer = split_buffer[-1]
+
+                return self.readline()
+
         b = conn.get_bucket(parsed_url.netloc, validate=False)
-        k = boto.s3.key.Key(b)
-        k.key = parsed_url.path
-        filepath_or_buffer = BytesIO(k.get_contents_as_string(
-            encoding=encoding))
+        if compat.PY2 and compression == "gzip":
+            k = boto.s3.key.Key(b, parsed_url.path)
+            filepath_or_buffer = BytesIO(k.get_contents_as_string(
+                encoding=encoding))
+        else:
+            k = OnceThroughKey(b, parsed_url.path, encoding=encoding)
```

Comment:
The function name means something like "open for reading" and is from from the `Key` class.  An `f.open().read()` would actually read some of the file content, which I don't want to do here. The main purpose of putting the `open_read` call here is to make sure that the file exists and is readable. Otherwise that wouldn't get caught until later. I can add a comment about the purpose.